prtpy/partitioning/integer_programming_avg.py

```diff
@@ -0,0 +1,151 @@
+    :param copies: how many copies there are of each item. Default: 1.
+    :param time_limit: stop the computation after this number of seconds have passed.
+
+    >>> from prtpy import BinnerKeepingContents, BinnerKeepingSums
+    >>> optimal(BinnerKeepingSums(), 2, [11.1,11,11,11,22],[0.1,0.9])
+    array([11. , 55.1])
+    >>> optimal(BinnerKeepingSums(), 2, [11.1,11,11,11,22],[0.9,0.1])
+    array([55.1, 11. ])
+    >>> optimal(BinnerKeepingSums(), 2, [11,11,11,11,22])
+    array([33., 33.])
+
+    >>> optimal(BinnerKeepingSums(), 5, [2,2,5,5,5,5,9])
+    array([5., 5., 7., 7., 9.])
+    >>> optimal(BinnerKeepingSums(), 3, [1,1,1,1])
+    array([1., 1., 2.])
+    >>> optimal(BinnerKeepingSums(), 3, [1,1,1,1,1])
+    array([1., 2., 2.])
+
+    The following examples are based on:
+        Walter (2013), 'Comparing the minimum completion times of two longest-first scheduling-heuristics'.
+    >>> walter_numbers = [46, 39, 27, 26, 16, 13, 10]
+    >>> printbins(optimal(BinnerKeepingContents(), 3, walter_numbers))
+    Bin #0: [39, 16], sum=55.0
+    Bin #1: [46, 13], sum=59.0
+    Bin #2: [27, 26, 10], sum=63.0
+
+    >>> printbins(optimal(BinnerKeepingContents(), 3, walter_numbers, [0.2, 0.5, 0.3]))
+    Bin #0: [26, 10], sum=36.0
+    Bin #1: [46, 27, 16], sum=89.0
+    Bin #2: [39, 13], sum=52.0
+
+    >>> printbins(optimal(BinnerKeepingContents(), 3, walter_numbers, [0.9, 0, 0.1]))
+    Bin #0: [46, 39, 27, 26, 13, 10], sum=161.0
+    Bin #1: [], sum=0.0
+    Bin #2: [16], sum=16.0
+
+    >>> optimal(BinnerKeepingSums(), 3, walter_numbers)
+    array([55., 59., 63.])
+
+    >>> from prtpy import partition
+    >>> partition(algorithm=optimal, numbins=3, items=[1, 2, 3], copies=[2, 1, 4])
+    [[2, 3], [1, 1, 3], [3, 3]]
+
+    """
+    ibins = range(numbins)
+    items = list(items)
+    iitems = range(len(items))
+    if isinstance(copies, Number):
+        copies = {iitem: copies for iitem in iitems}
+    model = mip.Model(name = '', solver_name=solver_name)
+    counts: dict = {
+        iitem: [model.add_var(var_type=mip.INTEGER) for ibin in ibins]
+        for iitem in iitems
+    }  # counts[i][j] is a variable that represents how many times item i appears in bin j.
+
+    bin_sums = [
+        sum([counts[iitem][ibin] * binner.valueof(items[iitem]) for iitem in iitems])
+        for ibin in ibins
+    ]
+
+    sum_items = 0
+    if not copies:
+        sum_items = sum(items)
+    else:
+        for i in range (len(items)):
+            sum_items = sum_items + items[i] * copies[i]
+
+    if relative_values:
+        z_js = [
+            bin_sums[ibin] - sum_items * relative_values[ibin]
+            for ibin in ibins
+        ]
+    else:
+        z_js = [
+            bin_sums[ibin] - sum_items / len(ibins)
+            for ibin in ibins
+        ]
+
+    t_js = [
+        model.add_var(var_type=mip.INTEGER) for ibin in ibins
+    ]
+
+
+    model.objective = mip.minimize(
+        0.5 * sum(t_js)
```

Comment:
Should it always be 0.5, regardless of the number of bins?

Reply:
Yes.
sum(t_js) calculates double the difference that we want because it calculates the positive and negative sums (using the absolute value). so we need to devide that into 2.